### ISSUE TITLE
f18 allows a UNIT specification CHARACTER PARAMETER but gfortran wants a CHARACTER variable

### ISSUE THREAD

**Issue by sscalpone** [flang-compiler/f18-llvm-project#360]
```
% cat rr.f90
      character,parameter :: str = "3E-1"
      read(str,*) r
      end
% flang -c rr.f90
% gfortran -c rr.f90
rr.f90:2:11:

    2 |       read(str,*) r
      |           1
Error: UNIT specification at (1) must be an INTEGER expression or a CHARACTER variable
```

**Comment by schweitzpgi** [comment]
Such an error message should be done in semantics.

**Comment by vdonaldson** [comment]
This involves a chain of syntax rules leading to constraint C901 -

 ```
 R1201 io-unit is file-unit-number
                or *
                or internal-file-variable

  R1203 internal-file-variable is char-variable

  R905 char-variable is variable

  R902 variable is designator
                or function-reference

  C901 (R902) designator shall not be a constant or a subobject of a constant.
```

Other chains of syntax rules also lead to constraint C901 -
 - IOSTAT= and IOMSG= specifiers
 - most INQUIRE specifiers
 - a few other specifiers

**Comment by vdonaldson** [comment]
Fixed in gitlab commit 2a0a8546ab20ae10aaa93d22064a7902fb3176fc

**Comment by sscalpone** [comment]
Fixed.  Verified.
